### PR TITLE
fix: move dependency serialization to an option, improve patching and unpatching (TC-49)

### PR DESCRIPTION
**Pull request by bcai2**
Adds:
- `serialize_dependencies` parameter (default value false)
- The ability to serialize a module and _all_ of its mentioned imports (deep serialization)

Deep serialization is disabled with no arg to enable, because I wasn't able to find a single example of something that needed deep serialization and didn't fail to pickle with `cloudpickle`.

### CODE REVIEW

File: src/lug/lug.py [hunk 1]
```diff
@@ -124,48 +149,65 @@ def patch_system_calls(func, user_docker_container_name, docker_shell_location):
...
 
-def unpatch_system_calls(func):
-    find_and_replace_function(
-        member=func.__globals__,
-        unique_docstring=None,
-        replace_with=None,
-        unpatch=True
-    )
+
+def unpatch_system_calls(patched_functions):
```
Comment (lebovic) [inline src/lug/lug.py:176]: This replaces the prior method of unpatching, which was calling `find_and_replace_function` with `unpatch=True`.

This keeps track of all patched functions, and then iterates through the tracked patched functions.

File: src/lug/lug.py [hunk 2]
```diff
@@ -124,48 +149,65 @@ def patch_system_calls(func, user_docker_container_name, docker_shell_location):
...
-        cloudpickle.register_pickle_by_value(inspect.getmodule(func))
-        cloudpickle.register_pickle_by_value(sys.modules[__name__])
-        pickled_patch = cloudpickle.dumps(patch_system_calls)
-        encoded_pickled_patch = base64.encodebytes(pickled_patch)
+        if serialize_dependencies:
+            modules_to_register = get_modules_to_register(func)
+            for module in modules_to_register:
+                cloudpickle.register_pickle_by_value(module)
+        else:
+            cloudpickle.register_pickle_by_value(sys.modules[__name__])
```
Comment (lebovic) [inline src/lug/lug.py:197]: Even if other dependencies aren't serialized, we still need `lug`

File: src/lug/module_detection.py [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+                if "_pytest" in module.__name__ or "url" in module.__name__:
+                    continue
+                deep_modules.add(sys.modules[module_name])
+        for direct_import in ast_direct_imports:
+            module_name = direct_import.names[0].name
+            if module_name in filtered_sys_modules and module_name not in modules_to_skip:
+                if "_pytest" in module.__name__ or "url" in module.__name__:
+                    continue
+                deep_modules.add(sys.modules[module_name])
+    return deep_modules
```
Comment (lebovic) [inline src/lug/module_detection.py:68]: The above finds _all_ imports within a dependency, with some limitations. In all tested cases, it causes pickling errors in the cases where it's needed.

That said, there's probably a case we didn't hit – and if by-value cloudpickle is improved, this would open new use-cases. As such, I'd prefer to keep it in place + tested, but not in active use.

Comment (bcai2) [reply src/lug/module_detection.py:68]: This part is hard to parse overall, though the variable names and individual lines are straightforward enough to follow. Thoughts on adding 1-line descriptor comments to "sections" of the code within this function, or is that an antipattern?

**Review by bcai2 — COMMENTED** [review]
1 comment, otherwise LGTM!

File: src/lug/module_detection.py (hunk 3 above)
Comment (lebovic) [reply src/lug/module_detection.py:68]: That's a good idea, will add!

Comment (lebovic) [reply src/lug/module_detection.py]: Added!

**Review by lebovic — APPROVED** [review]
